### PR TITLE
Let grace ticks to roll over into multiple leader slots

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
Each leader node gives previous leader some grace ticks to let it finish its leader slot. The grace ticks are calculated based on number of ticks in 1 slot. With smaller slots (400ms), the number of grace ticks has reduced. This will cause nodes in geographically distributed deployments (with higher latencies) to more frequently timeout previous leader slots (as latency might be > grace ticks). This will lead to forks that could have been avoided.

#### Summary of Changes
Each leader is scheduled for multiple slots. Instead of using 1 slot to calculate maximum grace ticks, use the total number of leader ticks to calculate it.

Fixes #

### CODE REVIEW

File: core/src/poh_recorder.rs [hunk 1]
```diff
@@ -21,12 +21,15 @@ use solana_sdk::poh_config::PohConfig;
 use solana_sdk::pubkey::Pubkey;
 use solana_sdk::timing;
 pub use solana_sdk::timing::Slot;
+use solana_sdk::timing::NUM_CONSECUTIVE_LEADER_SLOTS;
```
Comment (rob-solana) [inline core/src/poh_recorder.rs]: this needs to be in genesis...

File: core/src/poh_recorder.rs [hunk 2]
```diff
@@ -139,7 +142,7 @@ impl PohRecorder {
 
         if let Some(target_tick) = self.start_leader_at_tick {
             // we've reached target_tick OR poh was reset to run immediately
-            if next_tick >= target_tick || self.start_tick + self.grace_ticks == target_tick {
+            if next_tick >= target_tick || self.start_tick + self.grace_ticks >= target_tick {
```
Comment (rob-solana) [inline core/src/poh_recorder.rs]: why did you change this line?

File: core/src/poh_recorder.rs [hunk 3]
```diff
@@ -165,7 +168,7 @@ impl PohRecorder {
             .map(|slot| {
                 (
                     Some(slot * ticks_per_slot + grace_ticks),
-                    (slot + 1) * ticks_per_slot - 1,
+                    (slot + NUM_CONSECUTIVE_LEADER_SLOTS + 1) * ticks_per_slot - 1,
```
Comment (rob-solana) [inline core/src/poh_recorder.rs]: I'd really prefer for this to be dynamic, i.e. poh recorder looks to see how many future slots belong to this node, instead of hard-coding it with constants.

File: core/src/poh_recorder.rs (hunk 2 above)
Comment (pgarg66) [reply core/src/poh_recorder.rs]: :) I had the opposite question in this PR https://github.com/solana-labs/solana/pull/5127

**Comment by rob-solana** [comment]
this is interesting as an experiment, even if I have comments about the implementation.  any effect on the testnets with this change?

**Comment by pgarg66** [comment]
> this is interesting as an experiment, even if I have comments about the implementation. any effect on the testnets with this change?

I am bringing up a testnet with this change. I'll update here the perf changes I observe

File: core/src/poh_recorder.rs (hunk 2 above)
Comment (rob-solana) [reply core/src/poh_recorder.rs]: are we sorted, then?

Comment (pgarg66) [reply core/src/poh_recorder.rs]: Yes. Sounds good.

**Comment by pgarg66** [comment]
Ran the testnet with this PR. Here are the observations:

The testnet had CPU only nodes, distributed across US and Europe (same configration as TdS testnet)

The first run: This PR and client tx_count = 20000
The second run: This PR and client tx_count = 50000
The third run: Without this PR and client tx_count = 20000

The locktower is more stable with this PR (even with the higher tx_count)
image

The TPS is better with this PR
image

**Comment by sagar-solana** [comment]
Did you see any validator errors on your runs?

**Comment by pgarg66** [comment]
> Did you see any validator errors on your runs?

No.
